### ISSUE TITLE
Blazor Server keep-alive & reconnection issues

### ISSUE THREAD

**Issue by garrettlondon1** [dotnet/aspnetcore#48724]
### Is there an existing issue for this?

- No

### Is your feature request related to a problem? Please describe the problem.

SignalR connection has a Keep-Alive (default 15 seconds) which is the criteria for the reconnection modal appearing.

This means if a button is clicked on Web, and the keep-alive just triggered 1 second ago, when the client disconnects they're waiting 14 seconds for the reconnection modal.. 

No DOM updates can be communicated over SignalR and user can continue clicking buttons

The worst part of this is:
- **Dom events build up and are transmitted to the server on reconnection**

This is a very poor user experience for users who expect feedback from buttons/actions that cannot be communicated from the server to the client via the broken socket connection.
- There are additional consequences for the server side

The developer has no way of intercepting any client side event within this potential 14 seconds of UI interaction

The reconnection modal is fine, you can customize the appearance pretty nicely. It's just about the developer meeting the reconnection process when it happens, not 15 seconds after it happens

### Repro steps

- Create a Blazor Server app and run it
- Browse to the app with the browser dev tools open
- On the networking tab in the browser dev tools set the network throttling to offline
- Try to use the app

Expected result: The disconnected UI is displayed almost immediately prior to or in response to any UI interactions
Actual result: The disconnected UI isn't displayed and clicking on the UI elements does nothing for about 15sec

### Describe the solution you'd like

Have not thought about the implications of this, but depending on client side interactions, we could reset the keep-alive and ping on every `@onclick` handler (or any other condition) and throttle it every 2-3 seconds to prevent continuous pinging of the server if a user spam clicks?

It seems silly to prevent the .1% or less times that the user will get disconnected and experience UI hanging, but it is such a significant issue (if reconnect modal doesn't properly appear), that I feel it's warranted.

### Additional context

Yes state being lost, and ability to resume circuits is important, but I think this is a bigger fundamental problem. One experience of this will pretty much turn you off of the framework for any enterprise web app, and send you the SPA way.

**Comment by MackinnonBuck** [comment]
Thanks for reaching out! We've moved this issue to the backlog for now to gather additional feedback from the community about whether this is a change that should be made.

**Comment by surayya-MS** [comment]
Related https://github.com/dotnet/aspnetcore/issues/48675

**Comment by halter73** [comment]
> Have not thought about the implications of this, but depending on client side interactions, we could reset the keep-alive and ping on every @OnClick handler (or any other condition) and throttle it every 2-3 seconds to prevent continuous pinging of the server if a user spam clicks?

I'm pretty sure the SignalR Core clients already reset the keep-alive timeout every time they receive any message from the server, not just a keep-alive. I also think the server only bothers sending a keep alive if there are no other messages sent within the [KeepAliveInterval](https://learn.microsoft.com/en-us/aspnet/core/signalr/configuration?view=aspnetcore-7.0&tabs=dotnet#configure-server-options-6). Correct me if I'm wrong @BrennanConroy.

What happens if you lower the `KeepAliveInterval` and `ClientTimeoutInterval`?

**Comment by garrettlondon1** [comment]
@halter73 I don't think it's about the keep-alive on the server. If a client is disconnected from the server how will it receive the DOM update for the reconnection modal?

That's why it's really the client side keep-alive that matters, how often the client pings the server. Blazor Server you can specify the HubOptions in Startup for the server, and modify the blazor server.js startup to change on the client as seen in #48675.

I lowered the client side keep-alive to 1 second.. The Client Timeout interval is when the "Disconnected" modal appears, so I think it's fine to leave at any value really, since the main problem I'm trying to solve here is immediately show "Reconnecting" when client disconnects, not when the next keep-alive (15 seconds by default) is executed.

**Comment by garrettlondon1** [comment]
> https://learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/server?view=aspnetcore-7.0

> The [KeepAliveInterval](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.huboptions.keepaliveinterval#microsoft-aspnetcore-signalr-huboptions-keepaliveinterval) isn't directly related to the reconnection UI appearing. The Keep-Alive interval doesn't necessarily need to be changed. If the reconnection UI appearance issue is due to timeouts, the [ClientTimeoutInterval](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.huboptions.clienttimeoutinterval#microsoft-aspnetcore-signalr-huboptions-clienttimeoutinterval) and [HandshakeTimeout](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.huboptions.handshaketimeout#microsoft-aspnetcore-signalr-huboptions-handshaketimeout) can be increased and the Keep-Alive interval can remain the same. The important consideration is that if you change the Keep-Alive interval, make sure that the client timeout value is at least double the value of the Keep-Alive interval and that the Keep-Alive interval on the client matches the server setting.


Can the docs provide more clarity about how the "Keep-Alive isn't directly related to the reconnection UI appearing"?

In the following example, the [ClientTimeoutInterval](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.huboptions.clienttimeoutinterval#microsoft-aspnetcore-signalr-huboptions-clienttimeoutinterval) is increased to 60 seconds, and the [HandshakeTimeout](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.huboptions.handshaketimeout#microsoft-aspnetcore-signalr-huboptions-handshaketimeout) is increased to 30 seconds.

**Comment by alexyakunin** [comment]
I don't think this happens exactly as you describe - as far as I understand,
- Blazor reacts on websocket disconnect
- TCP keep-alive (as far as I understand, the KeepAlive is exactly this) is just one of ways to determine for TCP/IP stack that the connection is dead. Others include no ACK to regular packets, network connection failure, etc.

Your idea eith sending keep-alive to detect the disconnection faster seem meaningless, because TCP implies that any transmission requires ACK, not only keep-alive packet. And it's easy to conclude that such a transmission happens when you click a button - to deliver the event notification to server.

**Comment by alexyakunin** [comment]
I don't think this happens exactly as you describe - as far as I understand,

Blazor reacts to websocket disconnection
TCP keep-alive (as far as I understand, the KeepAlive is exactly this) is just one of ways to determine for TCP/IP stack that the connection is dead. Others include no ACK to regular packets, network connection failure, etc.
Your idea eith sending keep-alive to detect the disconnection faster seem meaningless, because TCP implies that any transmission requires ACK, not only keep-alive packet. And it's easy to conclude that such a transmission happens when you click a button - to deliver the event notification to server.

**Comment by alexyakunin** [comment]
What might be the real cause is SignalR reconnection - I am not sure about all the details, but it might be masking the fact that websocket connection is dead & it tries to reconnect while behaving like nothing happened.

**Comment by garrettlondon1** [comment]
I've been going through [src/Components/Web.JS/src/Platform/Circuits/DefaultReconnectionHandler.ts](https://github.com/dotnet/aspnetcore/blob/main/src/Components/Web.JS/src/Platform/Circuits/DefaultReconnectionHandler.ts) to try and get a better understanding.. but no luck. 

> The Keep-Alive interval (keepAliveIntervalInMilliseconds or [KeepAliveInterval](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.signalr.client.hubconnection.keepaliveinterval#microsoft-aspnetcore-signalr-client-hubconnection-keepaliveinterval)) **isn't directly related** to the reconnection UI appearing. The Keep-Alive interval **doesn't necessarily** need to be changed. If the reconnection UI appearance issue is due to timeouts, the server timeout can be increased and the Keep-Alive interval can remain the same. The important consideration is that if you change the Keep-Alive interval, make sure that the timeout value is at least double the value of the Keep-Alive interval and that the Keep-Alive interval on the server matches the client setting.

@danroth27 @MackinnonBuck @jongalloway Terms like "isn't directly related" or "doesn't necessarily need to be changed".. fairly confusing for technical documentation.

Imagine you're building a CRUD app and the websocket drops when your user clicks "Create sensitive record", nothing happens, no reconnection modal.. Your user will continue to click that button dozens of times and once the connection re-establishes, now you have flooded your API, and the user is even more frightened.

From a framework perspective server-side blazor needs to be able to offer the application developer the choice to **immediately** notify the user of the reconnecting process, or leave the default settings. Note: only have experience with Blazor Server deployed on App Service using Azure SignalR


@alexyakunin
> but it might be masking the fact that websocket connection is dead & it tries to reconnect while behaving like nothing happened.

 If what you are saying here is true: the websocket connection has severed, and it's trying in the background without triggering the Reconnecting modal, why delay the Reconnecting modal?

**Comment by danroth27** [comment]
> Terms like "isn't directly related" or "doesn't necessarily need to be changed".. fairly confusing for technical documentation.

@garrettlondon1 Agreed! To help get this fixed in the docs, could you please report an issue for the specific doc page by clicking the link at the bottom of the doc page?

![image](https://github.com/dotnet/aspnetcore/assets/1874516/f8cb0ae7-b9da-4c07-8fa5-5ca89603508c)

**Comment by garrettlondon1** [comment]
@danroth27 Much appreciated, I will take care of that right now.

**Comment by garrettlondon1** [comment]
Hey @mkArtakMSFT curious, why does this fall under the umbrella of Technical Debt? Even with .NET8, my understanding is that nothing has changed with the InteractiveServer components (Blazor Server), they rely on the same SignalR connection.

This issue is geared towards solving one of the biggest and **still current** issue of Blazor Server itself. Poor UI experience with socket disconnects.

**Comment by glen-84** [comment]
If there's no connectivity, it seems dangerous to record all events and replay them on reconnection.

I agree that the disconnected UI should display immediately when an action fails.

**Comment by garrettlondon1** [comment]
> If there's no connectivity, it seems dangerous to record all events and replay them on reconnection.
> 
> I agree that the disconnected UI should display immediately when an action fails.

Exactly @glen-84, very dangerous... For some applications, sure! Example: @captainsafia and teams AMAZING .NET8 demo with the websocket going down in the chat app "going out for lunch".

For an enterprise application which is important DB calls.. If a button state DOM change can't even be communicated over SignalR, now the user has no idea that if they click 20 times, 20 db calls will be queued when they reconnect.. And now you're forcing InteractiveServer component developers to implement some kind of rate limiting/serious error handling when they have no public API besides UI interactions?

Interactive Server components are not background websocket connections, where you can just use JS to handle SignalR disconnects. The whole thing **IS SIGNALR**!

**Comment by halter73** [comment]
We want to look into having the Blazor client validate that each non-keep-alive message it sends to the server gets and quick ack. The default could be as low as 2 seconds but should be configurable. This could probably be implemented using Hub method return values.

**Comment by garrettlondon1** [comment]
> We want to look into having the Blazor client validate that each non-keep-alive message it sends to the server gets and quick ack. The default could be as low as 2 seconds but should be configurable. This could probably be implemented using Hub method return values.

The UX behavior around this is also important. 

For example, on a Blazor Web App which is majority SSR, and one small InteractiveServer component.
- There needs to be a way to tell if that one "island" has lost interactivity. But **full page should not be as impacted since SSR could work with interactive technology outside of Websocket**

In a Blazor Web App that is mostly InteractiveServer based (whole pages, etc).. the traditional reconnection modal is better because if there is no interactivity, the UX is severely degraded as explained above.

**Comment by ProTip** [comment]
@garrettlondon1 I have worked with the implementation of a lot websocket clients(including Rails action cable, any cable, etc) and created a few myself.. Took a look at the code and I see a few things going on here.

First, _most_ disconnect events should result in WS disconnect events being raised, and the retry strategy should start immediately. This would include network changes, graceful server restarts, connection loses, and etc.

Unfortunately the reconnect strategy is **very naive**.  Instead of some variation of exponential backoff with full jitter, we see a that there is a first attempt max of 3s followed immediately by falling back to 20s which is the default retry interval in the options:
https://github.com/dotnet/aspnetcore/blob/v8.0.1/src/Components/Web.JS/src/Platform/Circuits/DefaultReconnectionHandler.ts#L69-L72

In my experience, which probably mirrors yours, it's best to do an effectively immediate retry(or maybe after 50ms) followed by some quick but increasing retries before falling back to an exponential backoff. So waiting 3s initially and then falling back to 20s is not a great UX.

I would like to see this strategy pluggable, but it seems there are some options for configuring this like so in `App.razor`:
```
<script src="_framework/blazor.web.js" autostart="false"></script>
    <script lang="javascript">
        Blazor.start({
            circuit: {
                reconnectionOptions: {
                    maxRetries: 100, retryIntervalMilliseconds: 500
                }
            }
        })
```

This will prevent Blazor from "auto" booting, then start it manually with reconnection options. It also appears an entire reconnection handler can be swapped in.

This doesn't address the situations where keepalive could help. Such as a communication disruption that doesn't result in a socket disconnect. However you may be able to detect this sooner via a global component that interacts with the server, and then by calling something like `window.Blazor.reconnect()` . I haven't investigated this thoroughly though.

**Comment by vgallegob** [comment]
> This issue is geared towards solving one of the biggest and **still current** issue of Blazor Server itself. Poor UI experience with socket disconnects.

I agree, the reconnects are very anoying. It would be nice to have more control on how the circuit gets shut down. 

For example, I have a fixed maximum number of circuits (say 500 sessions). I would like to be able to leave the circuit alive on the server till reconnected. But I would not want SignalR to buffer messages, I think this would be very dangerous.

In my case a user can not login in 2 devices at the same time. This means I could basically kill circuits coming from a old device when a user login from a new device. And I could keep the new device circuits alive indeterminately, or at least for some hours.

this is just some brainstorming on ways we could improve reconnects.

**Comment by Johnhersh** [comment]
Doesn't SignalR have a fallback mechanism to use just regular http requests if web sockets are not available?

**Comment by garrettlondon1** [comment]
@mkArtakMSFT I believe @danroth27 added the "Bug" label for a reason here:


### Repro steps
  Create a Blazor Server app and run it
  Browse to the app with the browser dev tools open
  On the networking tab in the browser dev tools set the network throttling to offline
  Try to use the app


This is a pretty basic bug I believe. When you disconnect, the reconnection modal does not appear.

**Comment by garrettlondon1** [comment]
It's kind of like if you were a kid at an arcade, put $1 in the game, press the button, nothing happens, so you naturally keep pressing the button.

Your intention was to only play once, now you've played 15 times and lost all your money and now you're crying because the game glitched and you've lost everything.
- That's Blazor Server in the enterprise without control over this keep-alive, just waiting to happen with sensitive financial orders, healthcare data, you name it.

**Comment by pradhankukiran** [comment]
I’ve been working on a fix for this based on @halter73’s ack-based detection idea.

Current approach:
  - Switch `beginInvokeDotNetFromJS`, `sendLocationChanged`, and `sendLocationChanging` in `CircuitManager.ts` from `send()` to `invoke<void>()`
  - Add a configurable `messageAckTimeoutMilliseconds` option (default `2000`, `0` disables the proactive timeout)
  - If the server doesn’t ack in time, or the `invoke()` rejects, force `connection.stop()` so the normal reconnect flow starts immediately
  - No server-side API or protocol changes: the existing `ComponentHub` methods already return `ValueTask`, and SignalR already emits a `CompletionMessage` for invocations
  with an `invocationId`
  - Scope is limited to user-interaction messages, not all circuit traffic

Known behavioral change:
  - `sendLocationChanged` / `sendLocationChanging` would now resolve on hub-method completion instead of local transport handoff, so they pick up about one RTT of latency.
  The hub methods return almost immediately, so this should be small, but I wanted to call it out explicitly.

  If this direction looks acceptable, I’m happy to open a PR.
